files/generate_system_metrics

```diff
@@ -0,0 +1,244 @@
+
+require "optparse"
+require "json"
+require "time"
+require "fileutils"
+
+# This script is intended to be run on a puppet infrastructure node as part of the
+# puppet_metrics_collector module.  It will generate system statistics in the file format that
+# other puppet_metrics_collector stats follow.  It will poll the system for data at a given
+# interval, and then output the average to a file once per given file interval.
+
+# General namespace for SystemMetrics module
+module SystemMetrics
+  # Main class for collecting system metrics into a json file
+  #
+  # @author Randell Pelak
+  #
+  # @attr [integer] polling_interval Time in seconds between calls to poll the system for data.
+  # @attr [integer] file_interval Time in seconds between the creation of each output file.
+  # @attr [string] metric_type cpu|memory
+  # @attr [string] metrics_dir The puppet_metrics_collector output directory.
+  # @attr [boolean] verbose Verbose output
+  # @attr [string] hostname Name of the host the metrics are from. In directory name and json file.
+  #
+  class GenerateSystemMetrics
+    #
+    # Initialize class
+    #
+    # @author Randell Pelak
+    #
+    # @param [integer] polling_interval Time in seconds between calls to poll the system for data.
+    # @param [integer] file_interval Time in seconds between the creation of each output file.
+    # @param [string] metric_type cpu|memory
+    # @param [string] metrics_dir The puppet_metrics_collector output directory.
+    # @param [boolean] verbose Verbose output
+    #
+    # @return [void]
+    def initialize(polling_interval, file_interval, metric_type, metrics_dir, verbose = false)
+      @polling_interval = polling_interval
+      @file_interval = file_interval
+      @metric_type = metric_type
+      @metrics_dir = metrics_dir
+      @verbose = verbose
+
+      @hostname = %x[hostname].strip
+      puts "Hostname is: #{@hostname}" if @verbose
+      FileUtils.mkdir_p(@metrics_dir) unless File.directory?(@metrics_dir)
+    end
+
+    # Run sar to collect the raw system data
+    #
+    # @author Randell Pelak
+    #
+    # @return [string] raw output from sar
+    def run_sar
+      times_to_poll = (@file_interval / @polling_interval).round
+      # sar inputs are polling interval and how many times to poll
+      comm_flags = " -r" if @metric_type =~ /memory/
+      comm = "sar #{comm_flags} #{@polling_interval} #{times_to_poll}"
+      puts "sar command is: #{comm}" if @verbose
+      %x[#{comm}]
+    end
+
+    # Parse the sar output and extract the metrics data
+    #
+    # @author Randell Pelak
+    #
+    # @param [array] sar_output
+    #
+    # @raise [RuntimeError] if sar_output doesn't parse correctly
+    #
+    # @return [hash] The metrics data
+    def parse_sar_output(sar_output)
+      sar_output_arr = sar_output.split(/\n+|\r+/).reject(&:empty?).map { |line| line.split }
+
+      headers_line = sar_output_arr.find { |e| e.include? "%user" }
```

Comment:
What is "%user"?  This is not included in my version of sar.

Reply:
This is what I see
[root@ip-10-227-0-89 ~]# sar 1 3
Linux 3.10.0-957.1.3.el7.x86_64 (ip-10-227-0-89.amz-dev.puppet.net) 	11/01/2019 	_x86_64_	(8 CPU)

04:47:43 PM     CPU     %user     %nice   %system   %iowait    %steal     %idle
04:47:44 PM     all      0.63      0.00      0.13      0.00      0.00     99.25
04:47:45 PM     all      0.12      0.00      0.00      0.00      0.00     99.88
04:47:46 PM     all      0.75      0.00      0.12      0.00      0.00     99.12
Average:        all      0.50      0.00      0.08      0.00      0.00     99.42
[root@ip-10-227-0-89 ~]#

the kbmemfree you posted before sounded like you were doing the mem version "sar -r 1 3"  But the point is good... cause the script supports both, and would have failed to find %user in the mem version... So doing your update for the reverse zip is critical.

Reply:
actually, the reverse thing doesn't solve the %user...  I will have to choose what to look for based on what metrics type is being run

Reply:
```
sar_output_arr.find { |e| e.include? "%user" or e.include? "kbmemfree" }
```

Reply:
I did it in a more wordy way. and tied it to @metric_type